### PR TITLE
Add Auto Interface Generation

### LINKED ISSUE #14: Add GOTCHA interfaces for MPI and HDF5
Enhance DFTracer with MPI and HDF5 Interfaces. Currently, the recorder does the following.
1. Extract function signatures manually from the header or online doc. 
2. Then use [Python script](https://github.com/uiuc-hpc/Recorder/tree/hdf5-pnetcdf-new-funcs/tools/generator) to generate code.
3. Those .txt files contain all function signatures, which may be useful for DFTracer

### PR DESCRIPTION
**Pull request by izzet**
Fixes #14 
- [x] Generate interfaces & capture functions via `llvm` given a header file (e.g. `hdf5.h`)
- [x] Make HDF5 optional and controlled through a build variable (e.g. `BRAHMA_BUILD_WITH_HDF5`)
- [x] Test HDF5 function capturing via `ctest` 

Installing dependencies for this task, either via Spack or `install_external_project`, is up for a debate.

### CODE REVIEW

**Comment by hariharan-devarajan** [comment]
@izzet it depends is this generation part of main build if not we can have an internal mechanism for generation

File: src/brahma/interceptor.cpp [hunk 1]
```diff
@@ -11,9 +11,14 @@ size_t total_apis = 0;
 extern int brahma_bind_functions() {
   BRAHMA_LOGGER_INIT();
   if (brahma::bindings == nullptr) {
-#ifdef BRAHMA_ENABLE_MPI
+#if defined(BRAHMA_ENABLE_HDF5) && defined(BRAHMA_ENABLE_MPI)
```
Comment (hariharan-devarajan) [inline src/brahma/interceptor.cpp:14]: Let's simplify this logic with 
Total = POSIX + stdio
Ifdef hdf

Total += hdf
Endif 

And so on.

File: test/test_hdf5.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
+//
+// Created by Izzet Yildirim <iyildirim@hawk.iit.edu> on 9/4/24.
+//
+#include <brahma/brahma.h>
+#ifdef BRAHMA_ENABLE_HDF5
+#include <hdf5.h>
+
+namespace brahma {
+class POSIXTest : public POSIX {
```
Comment (hariharan-devarajan) [inline test/test_hdf5.cpp:9]: If it is hdf5 specific do we need others?

File: test/test_posix_mpi.cpp [hunk 3]
```diff
@@ -1,9 +1,6 @@
 //
 // Created by hariharan on 8/8/22.
 //
-/* Brahma Header */
```
Comment (hariharan-devarajan) [inline test/test_posix_mpi.cpp:4]: Why did we remove the import segregation?

File: tools/generate_interfaces.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
+import argparse
```
Comment (hariharan-devarajan) [inline tools/generate_interfaces.py:1]: Let's add comments on this file for me really understand the logic

**Review by hariharan-devarajan — CHANGES_REQUESTED** [review]
Some minor changes.

**Comment by izzet** [comment]
> @izzet it depends is this generation part of main build if not we can have an internal mechanism for generation

No, auto-generation only works when `BRAHMA_GENERATE_INTERFACES=ON` and is not part of main compilation.

File: test/test_hdf5.cpp (hunk 2 above)
Comment (izzet) [reply test/test_hdf5.cpp:9]: It was printing some warning about POSIX being used but not captured (via GOTCHA?)

File: src/brahma/interceptor.cpp (hunk 1 above)
Comment (izzet) [reply src/brahma/interceptor.cpp:14]: Yeah, this makes sense 😄

File: test/test_hdf5.cpp (hunk 2 above)
Comment (hariharan-devarajan) [reply test/test_hdf5.cpp:9]: This is the part about conditional binding I was talking about. It's binding everything all the time. It should just be a warning and not an error. Can you find the source? I suggest we don't go through a warning; we can switch to debug.

**Comment by hariharan-devarajan** [comment]
> > @izzet it depends is this generation part of main build if not we can have an internal mechanism for generation
> 
> No, auto-generation only works when `BRAHMA_GENERATE_INTERFACES=ON` and is not part of main compilation.

Then, is it hard to have external project-based build dependencies for autogeneration? If it is simple, let's add that. Then, I would like to test this myself as well. Also, let's add different HDF5 and MPI versions to test this on CI. 

Check the version of spectrum mpi (which is openmpi) on lassen. Typically MPI are very stable so I am not worried as much. 

For HDF5 target 1.8, 1.12, and 1.14

**Comment by hariharan-devarajan** [comment]
@izzet I do not see my comments addressed. Once u do, please request a re-review.

**Comment by hariharan-devarajan** [comment]
Why did u close this?

**Pull request closed without merging**